_source/_posts/2021-08-02-fix-common-problems-cors.md

```diff
@@ -0,0 +1,530 @@
+author: phill-edwards
+by: contractor
+communities: [go]
+description: "Tutorial: This post walks through troubleshooting and fixing common problems associated with calling REST APIs from JavaScript."
+tags: [go, cors, security, javascript]
+tweets:
+- "Is your browser not letting you make requests to your REST API? It could be a CORS issue‼️"
+- "Have a REST API on a different domain from your web page. Your browser could block them, and here is why ⬇️"
+image: blog/fix-common-problems-cors/social.png
+type: awareness
+---
+
+Many websites have JavaScript functions that make network requests to a server, such as a REST API. The web pages and APIs are often in different domains. This introduces security issues in that any website can request data from an API. Cross-Origin Resource Sharing (CORS) provides a solution to these issues. It became a W3C recommendation in 2014. It makes it the responsibility of the web browser to prevent unauthorized access to APIs.  All modern web browsers enforce CORS. They prevent JavaScript from obtaining data from a server in a domain different than the domain the website was loaded from, unless the REST API server gives permission.
+
+From a developer's perspective, CORS is often a cause of much grief when it blocks network requests. CORS provides a number of different mechanisms for limiting JavaScript access to APIs. It is often not obvious which mechanism is blocking the request. We are going to build a simple web application that makes REST calls to a server in a different domain. We will deliberately make requests that the browser will block because of CORS policies and then show how to fix the issues. Let's get started!
+
+**NOTE**: The code for this project can be found on [GitHub](https://github.com/oktadev/okta-go-cors-example).
+
+## Prerequisites to Building a Go Application
+
+First things first, if you don't already have Go installed on your computer you will need to [download and install the Go Programming Language](https://golang.org/doc/install).
+
+Now, create a directory where all of our future code will live.
+
+```bash
+mkdir cors
+cd cors
+```
+
+Finally, we will make our directory a Go module and install the Gin package (a Go web framework) to implement a web server.
+
+```bash
+go mod init cors
+go get github.com/gin-gonic/gin
+go get github.com/gin-contrib/static
+```
+
+A file called `go.mod` will get created containing the dependencies.
+
+## How to Build a Simple Web Front End
+
+We are going to build a simple HTML and JavaScript front end and serve it from a web server written using Gin.
+
+First of all, create a directory called `frontend` and create a file called `frontend/index.html` with the following content:
+
+```html
+<html>
+    <head>
+        <meta charset="UTF-8" />
+        <title>Fixing Common Issues with CORS</title>
+        <script src="control.js" defer></script>
+    </head>
+    <body>
+        <h1>Fixing Common Issues with CORS</h1>
+        <div class="centred">
```

Comment:
Seems strange to have a class with no CSS.

Reply:
Good catch!